In this notebook, we will look at building regression models for decision tree, random forest and SVM.

We are going to start with Linear Regression on teh Car purchase dataset. Previously, we considered this dataset for linear regression. 

In [1]:
import pandas as pd
import numpy as np
from sklearn import linear_model
from sklearn.model_selection import train_test_split

In [2]:
auto = pd.read_csv('Car_Purchasing_Data.csv')
print(auto.head())

print(auto.shape)


auto.head()

print(auto.columns)
print(auto.dtypes)

print(auto.Country.unique())

     Customer Name                                    Customer e-mail Country  \
0    Martina Avila  cubilia.Curae.Phasellus@quisaccumsanconvallis.edu     USA   
1    Harlan Barnes                                eu.dolor@diam.co.uk     USA   
2  Naomi Rodriquez  vulputate.mauris.sagittis@ametconsectetueradip...     USA   
3  Jade Cunningham                            malesuada@dignissim.com     USA   
4     Cedric Leach     felis.ullamcorper.viverra@egetmollislectus.net     USA   

   Gender  Age  Annual Salary  Credit Card Debt    Net Worth  \
0       0   42    62812.09301      11609.380910  238961.2505   
1       0   41    66646.89292       9572.957136  530973.9078   
2       1   43    53798.55112      11160.355060  638467.1773   
3       1   58    79370.03798      14426.164850  548599.0524   
4       1   57    59729.15130       5358.712177  560304.0671   

   Car Purchase Amount  
0          35321.45877  
1          45115.52566  
2          42925.70921  
3          67422.36313  
4  

In [3]:
# dropping 

auto.drop(["Customer Name", "Customer e-mail", "Country"], axis=1, inplace=True)

#df.drop(['column_name1', 'column_name2'], axis=1, inplace=True)

print(auto.shape)

print(auto.columns)
print(auto.info())

(500, 6)
Index(['Gender', 'Age', 'Annual Salary', 'Credit Card Debt', 'Net Worth',
       'Car Purchase Amount'],
      dtype='object')
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 6 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Gender               500 non-null    int64  
 1   Age                  500 non-null    int64  
 2   Annual Salary        500 non-null    float64
 3   Credit Card Debt     500 non-null    float64
 4   Net Worth            500 non-null    float64
 5   Car Purchase Amount  500 non-null    float64
dtypes: float64(4), int64(2)
memory usage: 23.6 KB
None


In [4]:
# creating an instance of LinearRegression class
reg = linear_model.LinearRegression()
print(reg)

LinearRegression()


In [5]:
from sklearn.preprocessing import StandardScaler

In [6]:
a_scaler = StandardScaler() # creating an instance of Standard Scaler

auto_independent = a_scaler.fit_transform(auto[['Age','Annual Salary', 'Credit Card Debt', 'Net Worth']])

auto_target = auto["Car Purchase Amount"]

# using train_test_split(), we are splitting the data into training and test 
# x_train, x_test, y_train, y_test

x_train, x_test, y_train, y_test = train_test_split(auto_independent, auto_target, test_size=0.2, random_state=4)

### Linear Regression

In [7]:
# we have to do the fit on the training data
reg.fit(x_train, y_train)

# y = mx + b here m is the coefficient (or slope) of x 
# and b is the intercept

print(reg.coef_) # 
print(reg.intercept_)



[6698.49987763 6568.62278558   20.6676723  5020.44202563]
44203.966801967785


#### Linear Regression Equation

The relationship between the independent and dependent variables is 
'Age','Annual Salary', 'Credit Card Debt', 'Net Worth' independent variables

car purchase amount is the dependent variable

car_purchase = 6698.49* age + 6568.62 * annual_salary + 20.66 * credit_debt + 5050.44 * net_worth + 44203.96

### Decision Tree Regressor

In [8]:
from sklearn.model_selection import cross_val_score
from sklearn.tree import DecisionTreeRegressor

dt = DecisionTreeRegressor(random_state=0, criterion="absolute_error")
dt_fit = dt.fit(x_train, y_train)

dt_scores = cross_val_score(dt_fit, x_train, y_train, cv = 5)
print("mean cross validation score: {}".format(np.mean(dt_scores)))
print("score without cv: {}".format(dt_fit.score(x_train, y_train)))

# on the test 
from sklearn.metrics import r2_score
print(r2_score(y_test, dt_fit.predict(x_test)))
print(dt_fit.score(x_test, y_test))

mean cross validation score: 0.8396011747138079
score without cv: 1.0
0.8638359844306374
0.8638359844306374


#### Random Forest Regressor

In [ ]:
from sklearn.ensemble import RandomForestRegressor


regr_obj = RandomForestRegressor(max_depth=3, random_state=0)
regr_obj.fit(x_train, y_train)

rtree_score = regr_obj.score(x_train, y_train, sample_weight=None)
print(rtree_score)

### SVM Regressor

In [ ]:
from sklearn.svm import SVR

svm_reg = SVR(kernel="linear", C=1.0, epsilon=0.2)
svm_reg.fit(x_train, y_train)

svm_score = svm_reg.score(x_train, y_train, sample_weight=None)
print(svm_score)